### PR TITLE
FLUID-6178: Fix and test case for fluid.prettyPrintJSON bypass of synthetic properties

### PR DESCRIPTION
**Pull request by amb26**

### CODE REVIEW

**Comment by simonbates** [comment]
@amb26 I'm seeing a failure in `framework-tests/core/html/FluidDebugging-test.html`assertion 7 "Fluid Debugging JS Tests: fluid.prettyPrintJSON: Caught synthetic property":

- Expected: undefined

**Comment by amb26** [comment]
Thanks for catching this bumble, @simonbates . I decided on the one hand to refine the FluidDebuggingTests to, to start with, actually check a real synthetic property in order to trap the [Synthetic property] message, rather than one that was merely prototypal - and then to write a 2nd test which merely verifies that it makes some attempt to log a prototypal property since I think I'm not ready to commit to any particular behaviour here - the original test dated from a less competent era of testing and was pretty slapdash about the way it tried to conflate the two issues. Then, separately, I have fixed the implementation of fluid.values and fluid.keys to put them out of their long-standing misery reflected by the comment that used to stand by the utility fluid.makeFlatten - this has the side-effect of removing the side-effect whereby fluid.keys would actually evaluate values too - this means that we can return to the original implementation of fluid.prettyPrintJSON and have the node-based tests pass - see what you think - Cheers

File: src/framework/core/js/Fluid.js [hunk 1]
```diff
@@ -682,22 +682,23 @@ var fluid = fluid || fluid_3_0_0;
...
-        return function (obj) {
-            var togo = [];
-            fluid.each(obj, function (/* value, key */) {
-                togo.push(arguments[index]);
-            });
-            return togo;
-        };
-    };
-
     /** Return the keys in the supplied object as an array. Note that this will return keys found in the prototype chain as well as "own properties", unlike Object.keys() **/
```
[the remark this answers is not in the document]
Comment (amb26) [reply src/framework/core/js/Fluid.js]: Good point, added such tests

File: tests/framework-tests/core/js/FluidDebuggingTests.js [hunk 2]
```diff
@@ -50,10 +50,21 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
...
+        var synthetic = {};
+        Object.defineProperty(synthetic, "b", {
+            enumerable: true,
+            get: function () {
+                return 3;
+            }
+        });
 
         var renderedSynthetic = fluid.prettyPrintJSON(synthetic);
         jqUnit.assertTrue("Caught synthetic property", renderedSynthetic.indexOf("[Synthetic property]") !== -1);
```
[the remark this answers is not in the document]
Comment (amb26) [reply tests/framework-tests/core/js/FluidDebuggingTests.js]: Adjusted as suggested

File: tests/framework-tests/core/js/FluidDebuggingTests.js [hunk 3]
```diff
@@ -50,10 +50,21 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
...
-        var proto = { b: 3 };
-        Synthetic.prototype = proto;
-        var synthetic = new Synthetic();
+        function Derived() {}
+        var proto = { protoProperty: 3 };
+        Derived.prototype = proto;
+        var derived = new Derived();
+
+        var renderedDerived = fluid.prettyPrintJSON(derived);
+        jqUnit.assertTrue("Caught prototype property", renderedDerived.indexOf("protoProperty") !== -1);
```
[the remark this answers is not in the document]
Comment (amb26) [reply tests/framework-tests/core/js/FluidDebuggingTests.js]: I'm not sure I'm ready to commit myself here as to what the rendered RHS of such a property should be at this stage, merely that some attempt is made to detect that the property exists. There are arguments that the property value should simply be rendered as normal if it is not actually synthetic in the prototype parent, or that it should be flagged as it currently is to represent that it is not an "own" property, or perhaps both. I'd prefer to wait for more usage experience before we specify the behaviour one way or the other ...

**Comment by amb26** [comment]
Thanks for the great review, @simonbates - ready for another round

**Pull request merged**